<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Introduction &mdash; spfa 0.1 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> spfa
          </a>
              <div class="version">
                0.1.0
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">API:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../sofa.html">SOFA</a></li>
<li class="toctree-l1"><a class="reference internal" href="../plots.html">plots</a></li>
<li class="toctree-l1"><a class="reference internal" href="../utils.html">tools</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">spfa</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Introduction</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/notebooks/sc_multiome_example.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt .copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
.jp-RenderedHTMLCommon table,
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
.jp-RenderedHTMLCommon thead,
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
.jp-RenderedHTMLCommon tr,
.jp-RenderedHTMLCommon th,
.jp-RenderedHTMLCommon td,
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
.jp-RenderedHTMLCommon th,
div.rendered_html th {
  font-weight: bold;
}
.jp-RenderedHTMLCommon tbody tr:nth-child(odd),
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
.jp-RenderedHTMLCommon tbody tr:hover,
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># colab button
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># Install SOFA + dependencies
!pip install --quiet biosofa
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># download data
!wget https://datasets.cellxgene.cziscience.com/484dbc33-c7dc-4e5e-9954-7f2a1cc849bc.h5ad # RNA-Seq
!mv 484dbc33-c7dc-4e5e-9954-7f2a1cc849bc.h5ad rna.h5ad
!wget https://datasets.cellxgene.cziscience.com/fecbd715-66f5-48ae-8c39-51a76d7f1d3d.h5ad # ATAC-Seq
!mv fecbd715-66f5-48ae-8c39-51a76d7f1d3d.h5ad atac.h5ad
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
--2024-11-06 09:37:23--  https://datasets.cellxgene.cziscience.com/484dbc33-c7dc-4e5e-9954-7f2a1cc849bc.h5ad
Resolving datasets.cellxgene.cziscience.com (datasets.cellxgene.cziscience.com)... 18.172.112.45, 18.172.112.61, 18.172.112.87, ...
Connecting to datasets.cellxgene.cziscience.com (datasets.cellxgene.cziscience.com)|18.172.112.45|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 405964465 (387M) [binary/octet-stream]
Saving to: ‘484dbc33-c7dc-4e5e-9954-7f2a1cc849bc.h5ad’

-c7dc-4e5e-9954-7f2  24%[===&gt;                ]  95.62M  22.3MB/s    eta 17s    ^C
--2024-11-06 09:37:30--  https://datasets.cellxgene.cziscience.com/fecbd715-66f5-48ae-8c39-51a76d7f1d3d.h5ad
Resolving datasets.cellxgene.cziscience.com (datasets.cellxgene.cziscience.com)... 18.172.112.45, 18.172.112.61, 18.172.112.87, ...
Connecting to datasets.cellxgene.cziscience.com (datasets.cellxgene.cziscience.com)|18.172.112.45|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 906950703 (865M) [binary/octet-stream]
Saving to: ‘fecbd715-66f5-48ae-8c39-51a76d7f1d3d.h5ad’

    fecbd715-66f5-4   2%[                    ]  20.59M  12.2MB/s               ^C
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>import warnings
warnings.filterwarnings(&#39;ignore&#39;)
import sys
sys.path.append(&quot;/home/capraz/hubershare/SOFA&quot;)
import sofa
import torch
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import matplotlib
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from muon import MuData
from sklearn.manifold import TSNE
import matplotlib.patches as mpatches
import scanpy as sc
import anndata as ad
from anndata import AnnData
import muon
from matplotlib import colors as mp_colors
</pre></div>
</div>
</div>
<section id="Introduction">
<h1>Introduction<a class="headerlink" href="#Introduction" title="Permalink to this headline"></a></h1>
<p>In this notebook we will explore how <code class="docutils literal notranslate"><span class="pre">SOFA</span></code> can be used to analyze multi-omics data from the DepMap <a class="reference external" href="#1,#2,#3,#4,#5">[1,2,3,4,5]</a>. Here we give a brief introduction what the SOFA model does and what it can be used for. For a more detailed description please refer to our preprint: <a class="reference external" href="https://doi.org/10.1101/2024.10.10.617527">https://doi.org/10.1101/2024.10.10.617527</a></p>
<section id="The-SOFA-model">
<h2>The SOFA model<a class="headerlink" href="#The-SOFA-model" title="Permalink to this headline"></a></h2>
<p>Given a set of real-valued data matrices containing multi-omic measurements from overlapping samples (also called views), along with sample-level guiding variables that capture additional properties such as batches or mutational profiles, SOFA extracts an interpretable lower-dimensional data representation, consisting of a shared factor matrix and modality-specific loading matrices. The goal of these factors is to explain the major axes of variation in the data. SOFA explicitly assigns a subset
of factors to explain both the multi-omics data and the guiding variables (guided factors), while preserving another subset of factors exclusively for explaining the multi-omics data (unguided factors). Importantly, this feature allows the analyst to discern variation that is driven by known sources from novel, unexplained sources of variability.</p>
<section id="Interpretation-of-the-factors-(Z)">
<h3>Interpretation of the factors (Z)<a class="headerlink" href="#Interpretation-of-the-factors-(Z)" title="Permalink to this headline"></a></h3>
<p>Analogous to the interpretation of factors in PCA, SOFA factors ordinate samples along a zero-centered axis, where samples with opposing signs exhibit contrasting phenotypes along the inferred axis of variation, and the absolute value of the factor indicates the strength of the phenotype. Importantly, SOFA partitions the factors of the low-rank decomposition into guided and unguided factors: the guided factors are linked to specific guiding variables, while the unguided factors capture global,
yet unexplained, sources of variability in the data. The factor values can be used in downstream analysis tasks related to the samples, such as clustering or survival analysis. The factor values are called Z in SOFA.</p>
</section>
<section id="Interpretation-of-the-loading-weights-(W)">
<h3>Interpretation of the loading weights (W)<a class="headerlink" href="#Interpretation-of-the-loading-weights-(W)" title="Permalink to this headline"></a></h3>
<p>SOFA’s loading weights indicate the importance of each feature for its respective factor, thereby enabling the interpretation of SOFA factors. Loading weights close to zero indicate that a feature has little to no importance for the respective factor, while large magnitudes suggest strong relevance. The sign of the loading weight aligns with its corresponding factor, meaning that positive loading weights indicate higher feature levels in samples with positive factor values, and negative loading
weights indicate higher feature levels in samples with negative factor values. The top loading weights can be simply inspected or used in downstream analysis such as gene set enrichment analysis. The factor values are called W in SOFA.</p>
</section>
<section id="Supported-data">
<h3>Supported data<a class="headerlink" href="#Supported-data" title="Permalink to this headline"></a></h3>
<p>SOFA expects a set of matrices containing omics measurements with matching and aligned samples and different features. Currently SOFA only supports Gaussian likelihoods, for the multi-omics data. Data should therefore be appropriately normalized according to its omics modality. Additionally, data should be centered and scaled.</p>
<p>For the guiding variables SOFA supports Gaussian, Bernoulli and Categorical likelihoods. Guiding variables can therefore be continuous, binary or categorical. Guiding variables should be vectors with matching samples with the multi-omics data.</p>
<p>In SOFA the multi-omics data is denoted as X and the guiding variables as Y.</p>
</section>
</section>
<section id="Single-cell-multiome-data-set-of-the-human-cortex">
<h2>Single-cell multiome data set of the human cortex<a class="headerlink" href="#Single-cell-multiome-data-set-of-the-human-cortex" title="Permalink to this headline"></a></h2>
<p>The data we analyze in this notebook was generated by <a class="reference external" href="#1">[1]</a>. The authors simultaneously profiled the transcriptome (RNA) and chromatin accessibility (ATAC) of 45549 single cells of the human cerebral cortex at 6 different developmental stages. The authors identified 13 different cell types in the data. We will fit a SOFA model with 15 factors and guide the first 13 factors with a different cell type label. The 13 guided factors will explain the molecular differences between the cell
types, while the 2 unguided factors are free to explain within cell type variation. We will first load the data and do some basic preprocessing, then fit a SOFA model and perform various downstream analyses.</p>
<p>[1] Zhu, K. et al. Multi-omic profiling of the developing human cerebral cortex at the single-cell level. Sci Adv 9, eadg3754 (2023).</p>
</section>
</section>
<section id="Load-and-preprocess">
<h1>Load and preprocess<a class="headerlink" href="#Load-and-preprocess" title="Permalink to this headline"></a></h1>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata_rna= sc.read_h5ad(&quot;/g/huber/users/capraz/supervised_matrix_factorization/brain_10x_multiome/rna.h5ad&quot;) #&quot;rna.h5ad&quot;)
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata_rna
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 45549 × 30113
    obs: &#39;author_cell_type&#39;, &#39;age_group&#39;, &#39;donor_id&#39;, &#39;nCount_RNA&#39;, &#39;nFeature_RNA&#39;, &#39;nCount_ATAC&#39;, &#39;nFeature_ATAC&#39;, &#39;TSS_percentile&#39;, &#39;nucleosome_signal&#39;, &#39;percent_mt&#39;, &#39;assay_ontology_term_id&#39;, &#39;cell_type_ontology_term_id&#39;, &#39;development_stage_ontology_term_id&#39;, &#39;disease_ontology_term_id&#39;, &#39;self_reported_ethnicity_ontology_term_id&#39;, &#39;organism_ontology_term_id&#39;, &#39;sex_ontology_term_id&#39;, &#39;tissue_ontology_term_id&#39;, &#39;suspension_type&#39;, &#39;is_primary_data&#39;, &#39;batch&#39;, &#39;tissue_type&#39;, &#39;cell_type&#39;, &#39;assay&#39;, &#39;disease&#39;, &#39;organism&#39;, &#39;sex&#39;, &#39;tissue&#39;, &#39;self_reported_ethnicity&#39;, &#39;development_stage&#39;, &#39;observation_joinid&#39;
    var: &#39;feature_is_filtered&#39;, &#39;feature_name&#39;, &#39;feature_reference&#39;, &#39;feature_biotype&#39;, &#39;feature_length&#39;
    uns: &#39;batch_condition&#39;, &#39;citation&#39;, &#39;schema_reference&#39;, &#39;schema_version&#39;, &#39;title&#39;
    obsm: &#39;X_joint_wnn_umap&#39;, &#39;X_umap&#39;
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># basic preprocessing
adata_rna.X = adata_rna.raw.X
# normalization to total library size
sc.pp.normalize_total(adata_rna)
# log transformation
sc.pp.log1p(adata_rna)
<br/></pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata_atac= sc.read_h5ad(&quot;/g/huber/users/capraz/supervised_matrix_factorization/brain_10x_multiome/atac.h5ad&quot;)
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># select highly variable genes
sc.pp.highly_variable_genes(
    adata_rna,
    n_top_genes=2000,
    flavor=&quot;seurat&quot;,
    subset=True
)
sc.pp.highly_variable_genes(
    adata_atac,
    n_top_genes=2000,
    flavor=&quot;seurat&quot;,
    subset=True
)
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># scale the data
sc.pp.scale(adata_rna)
sc.pp.scale(adata_atac)
</pre></div>
</div>
</div>
<section id="Set-up-Xmdata-for-SOFA">
<h2>Set up Xmdata for SOFA<a class="headerlink" href="#Set-up-Xmdata-for-SOFA" title="Permalink to this headline"></a></h2>
<section id="Manually">
<h3>Manually<a class="headerlink" href="#Manually" title="Permalink to this headline"></a></h3>
<p>SOFA requires the following slots in uns: * llh: “gaussian” Currently only the Gaussian likelihood for the multi-omics data is supported.</p>
<p>and in obsm: * mask: boolean vector of length number of samples that masks samples with missing values</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>metadata = adata_rna.obs
adata_rna.uns[&quot;llh&quot;] = &quot;gaussian&quot;
adata_rna.X = adata_rna.X
adata_rna.obsm[&quot;mask&quot;] = ~np.any(np.isnan(adata_rna.X), axis=1)
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata_atac.uns[&quot;llh&quot;] = &quot;gaussian&quot;
adata_atac.X = adata_atac.X
adata_atac.obsm[&quot;mask&quot;] = ~np.any(np.isnan(adata_atac.X), axis=1)

adata_rna.var_names = adata_rna.var[&quot;feature_name&quot;]
adata_atac.var_names = adata_atac.var[&quot;feature_name&quot;]
</pre></div>
</div>
</div>
</section>
<section id="Using-SOFA’s-sofa.tl.get_ad()">
<h3>Using SOFA’s sofa.tl.get_ad()<a class="headerlink" href="#Using-SOFA’s-sofa.tl.get_ad()" title="Permalink to this headline"></a></h3>
<p>Alternatively we can use SOFA’s inbuilt sofa.tl.get_ad() function to create the appropriate <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> object</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># First we convert the `AnnData` objects to dataframes
rna_df = adata_rna.to_df()
atac_df = adata_atac.to_df()
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># Then use sofa.tl.get_ad() with default parameters
adata_atac = sofa.tl.get_ad(rna_df)
adata_rna = sofa.tl.get_ad(atac_df)
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># wrap the individual `AnnData` objects in a `MuData`
Xmdata = MuData({&quot;RNA&quot;:adata_rna, &quot;ATAC&quot;:adata_atac})
Xmdata
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<pre>MuData object with n_obs × n_vars = 45549 × 4000
  2 modalities
    RNA:    45549 x 2000
      uns:  &#x27;llh&#x27;, &#x27;scaling_factor&#x27;
      obsm: &#x27;mask&#x27;
    ATAC:   45549 x 2000
      uns:  &#x27;llh&#x27;, &#x27;scaling_factor&#x27;
      obsm: &#x27;mask&#x27;</pre></div>
</div>
</section>
</section>
<section id="Set-up-Ymdata-for-SOFA">
<h2>Set up Ymdata for SOFA<a class="headerlink" href="#Set-up-Ymdata-for-SOFA" title="Permalink to this headline"></a></h2>
<p>As described in the introduction, we would like to guide the first 13 factors with the 13 cell type labels. To this end we first need to one hot encode the cell type labels. We make use of the OneHotEncoder of <code class="docutils literal notranslate"><span class="pre">scikit-learn</span></code>:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>onc = OneHotEncoder()
onc_data = np.array(onc.fit_transform(metadata[[&quot;cell_type&quot;]]).todense())
# convert to df for later
celltype_df = pd.DataFrame(onc_data, columns= onc.categories_[0])
celltype_df
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>astrocyte</th>
      <th>caudal ganglionic eminence derived interneuron</th>
      <th>endothelial cell</th>
      <th>glutamatergic neuron</th>
      <th>inhibitory interneuron</th>
      <th>medial ganglionic eminence derived interneuron</th>
      <th>microglial cell</th>
      <th>neural progenitor cell</th>
      <th>oligodendrocyte</th>
      <th>oligodendrocyte precursor cell</th>
      <th>pericyte</th>
      <th>radial glial cell</th>
      <th>vascular associated smooth muscle cell</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>3</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>45544</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>45545</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>45546</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>45547</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>45548</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
  </tbody>
</table>
<p>45549 rows × 13 columns</p>
</div></div>
</div>
<p>each columns represents a different one hot (binary) encoded cell type</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># convert each column to an `AnnData` object using sofa.tl.get_ad() and store in a list
# as each cell type is now represented as a binary vector we need to choose the &quot;bernoulli&quot; likelihood
celltype_ad = [sofa.tl.get_ad(pd.DataFrame(celltype_df.iloc[:,i], columns=[str(onc.categories_[0][i])]), llh = &quot;bernoulli&quot;) for i in range(celltype_df.shape[1])]
# convert the list to a dictionary
cell_type_dict = {str(onc.categories_[0][i]):celltype_ad[i] for i in range(celltype_df.shape[1])}
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># wrap dictionary as `MuData`
Ymdata = MuData(cell_type_dict)
Ymdata
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<pre>MuData object with n_obs × n_vars = 45549 × 13
  13 modalities
    astrocyte:      45549 x 1
      uns:  &#x27;llh&#x27;, &#x27;scaling_factor&#x27;
      obsm: &#x27;mask&#x27;
    caudal ganglionic eminence derived interneuron: 45549 x 1
      uns:  &#x27;llh&#x27;, &#x27;scaling_factor&#x27;
      obsm: &#x27;mask&#x27;
    endothelial cell:       45549 x 1
      uns:  &#x27;llh&#x27;, &#x27;scaling_factor&#x27;
      obsm: &#x27;mask&#x27;
    glutamatergic neuron:   45549 x 1
      uns:  &#x27;llh&#x27;, &#x27;scaling_factor&#x27;
      obsm: &#x27;mask&#x27;
    inhibitory interneuron: 45549 x 1
      uns:  &#x27;llh&#x27;, &#x27;scaling_factor&#x27;
      obsm: &#x27;mask&#x27;
    medial ganglionic eminence derived interneuron: 45549 x 1
      uns:  &#x27;llh&#x27;, &#x27;scaling_factor&#x27;
      obsm: &#x27;mask&#x27;
    microglial cell:        45549 x 1
      uns:  &#x27;llh&#x27;, &#x27;scaling_factor&#x27;
      obsm: &#x27;mask&#x27;
    neural progenitor cell: 45549 x 1
      uns:  &#x27;llh&#x27;, &#x27;scaling_factor&#x27;
      obsm: &#x27;mask&#x27;
    oligodendrocyte:        45549 x 1
      uns:  &#x27;llh&#x27;, &#x27;scaling_factor&#x27;
      obsm: &#x27;mask&#x27;
    oligodendrocyte precursor cell: 45549 x 1
      uns:  &#x27;llh&#x27;, &#x27;scaling_factor&#x27;
      obsm: &#x27;mask&#x27;
    pericyte:       45549 x 1
      uns:  &#x27;llh&#x27;, &#x27;scaling_factor&#x27;
      obsm: &#x27;mask&#x27;
    radial glial cell:      45549 x 1
      uns:  &#x27;llh&#x27;, &#x27;scaling_factor&#x27;
      obsm: &#x27;mask&#x27;
    vascular associated smooth muscle cell: 45549 x 1
      uns:  &#x27;llh&#x27;, &#x27;scaling_factor&#x27;
      obsm: &#x27;mask&#x27;</pre></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>num_factors = 15
# In order to relate factors to guiding variables we need to provide a design matrix (guiding variables x number of factors)
# indicating which factor is guided by which guiding variable.
# Here we just indicate that the first 6 factors are each guided by a different guiding variable:
design = np.zeros((len(Ymdata.mod), num_factors))
for i in range(len(Ymdata.mod)):
    design[i,i] = 1
# convert to torch tensor to make it usable by SOFA
design = torch.tensor(design)
design
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
tensor([[1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0.]],
       dtype=torch.float64)
</pre></div></div>
</div>
</section>
</section>
<section id="Model">
<h1>Model<a class="headerlink" href="#Model" title="Permalink to this headline"></a></h1>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[93]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>model = sofa.SOFA(Xmdata=Xmdata,
             num_factors=num_factors,
             Ymdata = Ymdata,
             design = design,
             device=&#39;cuda&#39;,
             subsample=2048, # for single-cell data it can be beneficial to subsample minibatches (here of size 2048) of the data for training, this speeds up the fitting process
             seed=42)
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[94]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>model.fit(n_steps=6000, lr=0.01, predict = True)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Current Elbo 2.50E+08 | Delta: -1523794: 100%|██████████| 6000/6000 [14:14&lt;00:00,  7.02it/s]
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># if we would like to save the fitted model we can save it using:
#sofa.tl.save_model(model,&quot;brain_example_model&quot;)

# to load the model use:
#model = sofa.tl.load_model(&quot;brain_example_model&quot;)
</pre></div>
</div>
</div>
</section>
<section id="Downstream-analysis">
<h1>Downstream analysis<a class="headerlink" href="#Downstream-analysis" title="Permalink to this headline"></a></h1>
<section id="Convergence">
<h2>Convergence<a class="headerlink" href="#Convergence" title="Permalink to this headline"></a></h2>
<p>We will first assess whether the ELBO loss of SOFA has converged by plotting it over training steps</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>plt.semilogy(model.history)
plt.xlabel(&quot;Training steps&quot;)
plt.ylabel(&quot;ELBO&quot;)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Text(0, 0.5, &#39;ELBO&#39;)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_sc_multiome_example_31_1.png" src="../_images/notebooks_sc_multiome_example_31_1.png" />
</div>
</div>
</section>
<section id="Variance-explained">
<h2>Variance explained<a class="headerlink" href="#Variance-explained" title="Permalink to this headline"></a></h2>
<p>A good first step in a SOFA analysis is to plot how much variance is explained by each factor for each modality. This gives us an overview which factors are active across multiple modalities, capturing correlated variation across multiple measurements and which are private to a single modality, most probably capturing technical effects related to this modality.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sofa.pl.plot_variance_explained(model)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Axes: xlabel=&#39;View&#39;, ylabel=&#39;Factor&#39;&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_sc_multiome_example_33_1.png" src="../_images/notebooks_sc_multiome_example_33_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># We can also plot how much variance of each view is explained
sofa.pl.plot_variance_explained_view(model)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Axes: xlabel=&#39;View&#39;, ylabel=&#39;R2&#39;&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_sc_multiome_example_34_1.png" src="../_images/notebooks_sc_multiome_example_34_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># or how much variance is explained by each factor in total
sofa.pl.plot_variance_explained_factor(model)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Axes: xlabel=&#39;View&#39;, ylabel=&#39;R2&#39;&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_sc_multiome_example_35_1.png" src="../_images/notebooks_sc_multiome_example_35_1.png" />
</div>
</div>
</section>
<section id="Check-factor-guidance">
<h2>Check factor guidance<a class="headerlink" href="#Check-factor-guidance" title="Permalink to this headline"></a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># plot correlation of factor values with cell type labels
sofa.pl.plot_factor_metadata_cor(model, celltype_df)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Axes: xlabel=&#39;Covariate&#39;, ylabel=&#39;Factor&#39;&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_sc_multiome_example_37_1.png" src="../_images/notebooks_sc_multiome_example_37_1.png" />
</div>
</div>
</section>
<section id="Downstream-analysis-of-the-factor-values">
<h2>Downstream analysis of the factor values<a class="headerlink" href="#Downstream-analysis-of-the-factor-values" title="Permalink to this headline"></a></h2>
<p>The factor values represent the new coordinates in lower dimensional space of our samples and have dimensions samples x factors. The factor values called Z in SOFA. We can use the factor values for all kinds of downstream analyses on the sample level. Here we will cluster the unguided factors.</p>
<p>We first retrieve the factor values:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>Z = sofa.tl.get_factors(model)
Z
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Factor_1 (astrocyte)</th>
      <th>Factor_2 (caudal ganglionic eminence derived interneuron)</th>
      <th>Factor_3 (endothelial cell)</th>
      <th>Factor_4 (glutamatergic neuron)</th>
      <th>Factor_5 (inhibitory interneuron)</th>
      <th>Factor_6 (medial ganglionic eminence derived interneuron)</th>
      <th>Factor_7 (microglial cell)</th>
      <th>Factor_8 (neural progenitor cell)</th>
      <th>Factor_9 (oligodendrocyte)</th>
      <th>Factor_10 (oligodendrocyte precursor cell)</th>
      <th>Factor_11 (pericyte)</th>
      <th>Factor_12 (radial glial cell)</th>
      <th>Factor_13 (vascular associated smooth muscle cell)</th>
      <th>Factor_14</th>
      <th>Factor_15</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>4_AAACAGCCAACACTTG-1</th>
      <td>0.611896</td>
      <td>-0.506728</td>
      <td>0.283791</td>
      <td>1.593037</td>
      <td>0.168398</td>
      <td>0.399242</td>
      <td>0.422193</td>
      <td>0.024536</td>
      <td>-0.414130</td>
      <td>-0.239863</td>
      <td>-0.121763</td>
      <td>0.381870</td>
      <td>0.133415</td>
      <td>-0.268314</td>
      <td>-0.236205</td>
    </tr>
    <tr>
      <th>4_AAACAGCCACCAAAGG-1</th>
      <td>0.369894</td>
      <td>-0.695359</td>
      <td>0.196381</td>
      <td>1.779255</td>
      <td>0.204215</td>
      <td>0.662641</td>
      <td>0.385849</td>
      <td>-0.214301</td>
      <td>-0.265921</td>
      <td>-0.233039</td>
      <td>0.073905</td>
      <td>0.207890</td>
      <td>-0.089338</td>
      <td>-0.538529</td>
      <td>-0.470559</td>
    </tr>
    <tr>
      <th>4_AAACAGCCATAAGTTC-1</th>
      <td>0.525656</td>
      <td>-0.576934</td>
      <td>0.151626</td>
      <td>2.239841</td>
      <td>0.164921</td>
      <td>0.566175</td>
      <td>0.381240</td>
      <td>-0.492999</td>
      <td>-0.334595</td>
      <td>-0.270106</td>
      <td>-0.003286</td>
      <td>0.406507</td>
      <td>-0.017237</td>
      <td>0.213183</td>
      <td>-0.319585</td>
    </tr>
    <tr>
      <th>4_AAACATGCATAGTCAT-1</th>
      <td>0.484234</td>
      <td>-0.433456</td>
      <td>0.094686</td>
      <td>1.731916</td>
      <td>0.098722</td>
      <td>0.500609</td>
      <td>0.417679</td>
      <td>-0.256506</td>
      <td>-0.528553</td>
      <td>-0.414691</td>
      <td>-0.151233</td>
      <td>0.122768</td>
      <td>-0.096484</td>
      <td>-0.026044</td>
      <td>-0.005257</td>
    </tr>
    <tr>
      <th>4_AAACATGCATTGTCAG-1</th>
      <td>0.473560</td>
      <td>-0.469949</td>
      <td>0.194220</td>
      <td>2.000594</td>
      <td>0.305656</td>
      <td>0.425527</td>
      <td>0.379652</td>
      <td>-0.437305</td>
      <td>-0.143152</td>
      <td>-0.080377</td>
      <td>0.052171</td>
      <td>0.357325</td>
      <td>-0.041771</td>
      <td>-0.080069</td>
      <td>-0.247479</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>150666_TTTGTGAAGACAACAG-1</th>
      <td>0.338253</td>
      <td>0.274610</td>
      <td>0.021684</td>
      <td>-0.642276</td>
      <td>0.297069</td>
      <td>-0.107472</td>
      <td>0.752773</td>
      <td>-0.331871</td>
      <td>1.272615</td>
      <td>-0.831026</td>
      <td>-1.536213</td>
      <td>0.170395</td>
      <td>1.086244</td>
      <td>0.420269</td>
      <td>1.975964</td>
    </tr>
    <tr>
      <th>150666_TTTGTGAAGGCTGTGC-1</th>
      <td>0.088728</td>
      <td>0.120536</td>
      <td>0.196474</td>
      <td>-0.235622</td>
      <td>0.401305</td>
      <td>0.021225</td>
      <td>0.439250</td>
      <td>-0.194579</td>
      <td>-1.017107</td>
      <td>1.676835</td>
      <td>-1.030105</td>
      <td>0.552844</td>
      <td>0.437842</td>
      <td>0.196519</td>
      <td>1.003077</td>
    </tr>
    <tr>
      <th>150666_TTTGTGAAGTAAGAAC-1</th>
      <td>0.254422</td>
      <td>-0.270921</td>
      <td>0.105018</td>
      <td>-0.476804</td>
      <td>0.007705</td>
      <td>0.118286</td>
      <td>0.191205</td>
      <td>-0.072784</td>
      <td>1.845788</td>
      <td>-0.173360</td>
      <td>-0.375505</td>
      <td>0.115170</td>
      <td>0.340696</td>
      <td>-0.138671</td>
      <td>0.186792</td>
    </tr>
    <tr>
      <th>150666_TTTGTGAAGTCTTGAA-1</th>
      <td>0.707852</td>
      <td>-0.214515</td>
      <td>0.089128</td>
      <td>-0.605852</td>
      <td>0.340981</td>
      <td>0.148991</td>
      <td>0.792436</td>
      <td>-0.148444</td>
      <td>2.024130</td>
      <td>-0.886393</td>
      <td>-1.443925</td>
      <td>0.235342</td>
      <td>0.654023</td>
      <td>0.699232</td>
      <td>1.188664</td>
    </tr>
    <tr>
      <th>150666_TTTGTTGGTGATCAGC-1</th>
      <td>0.890338</td>
      <td>-0.290386</td>
      <td>0.247877</td>
      <td>-0.622976</td>
      <td>0.332508</td>
      <td>0.064684</td>
      <td>1.110440</td>
      <td>-0.020911</td>
      <td>2.552641</td>
      <td>-1.316348</td>
      <td>-1.953289</td>
      <td>0.298762</td>
      <td>1.031413</td>
      <td>1.323803</td>
      <td>1.993683</td>
    </tr>
  </tbody>
</table>
<p>45549 rows × 15 columns</p>
</div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># here we add the factors Z as obsm to the input object to be able to use scanpy for UMAP visualization
Xmdata.mod[&quot;RNA&quot;].obsm[&quot;Z&quot;] = Z.values
sc.pp.neighbors(Xmdata.mod[&quot;RNA&quot;], use_rep=&quot;Z&quot;, key_added=&quot;Z_neighbors&quot;)
sc.tl.umap(Xmdata.mod[&quot;RNA&quot;], neighbors_key=&quot;Z_neighbors&quot;, random_state=1)
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># plot the UMAP
Xmdata.mod[&quot;RNA&quot;].obs = metadata
sc.pl.embedding(Xmdata.mod[&quot;RNA&quot;], basis = &quot;X_umap&quot;, color=&quot;cell_type&quot;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_sc_multiome_example_41_0.png" src="../_images/notebooks_sc_multiome_example_41_0.png" />
</div>
</div>
<p>As expected the cell type separate perfectly in the UMAP visualization. This is expected as we guided the first 13 factors with the cell type labels. We can now overlay the unguided factors and check if we can identify interesting patterns:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># add each factor as obs to the input object
Xmdata.mod[&quot;RNA&quot;].obs = pd.concat((Xmdata.mod[&quot;RNA&quot;].obs, Z), axis=1)
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[38]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># plot a UMAP for each factor and color by factor value
fig,axes = plt.subplots(nrows =3, ncols=5,figsize=(34,20))

axes = axes.flatten()
for i,ax in enumerate(axes):

        divnorm=mp_colors.TwoSlopeNorm(vmin=np.min(Z.iloc[:,i]), vcenter=0, vmax=np.max(Z.iloc[:,i]))
        plot = ax.scatter(Xmdata.mod[&quot;RNA&quot;].obsm[&quot;X_umap&quot;][:,0], Xmdata.mod[&quot;RNA&quot;].obsm[&quot;X_umap&quot;][:,1], c=Z.iloc[:,i], s=120000 / Xmdata.mod[&quot;RNA&quot;].shape[0], cmap=&quot;RdBu&quot;, norm=divnorm)
        ax.set_aspect(&quot;equal&quot;)
        if model.Ymdata is not None and i &lt; len(model.Ymdata.mod):
            covariate = list(model.Ymdata.mod.keys())[i]
            ax.set_title(&quot;Factor &quot; +str(i+1) + &quot; (&quot;+ str(covariate))
        else:
            ax.set_title(&quot;Factor &quot; +str(i+1))
        plt.colorbar(plot)
        ax.set_xlabel(&quot;UMAP 1&quot;)
        ax.set_ylabel(&quot;UMAP 2&quot;)
        ax.tick_params(
        axis=&#39;both&#39;,          # changes apply to the x-axis
        which=&#39;both&#39;,      # both major and minor ticks are affected
        bottom=False,      # ticks along the bottom edge are off
        top=False,
        left=False,# ticks along the top edge are off
        labelbottom=False,
    labelleft=False) # labels along the bottom edge are off

plt.tight_layout()
matplotlib.rcdefaults()
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_sc_multiome_example_44_0.png" src="../_images/notebooks_sc_multiome_example_44_0.png" />
</div>
</div>
<p>While the guided factors are active in their respective cell type clusters, the unguided factors (14 and 15) show gradients within in cell type clusters. Factor 14 shows a gradient mostly active in the microglial cell cluster.</p>
</section>
<section id="Downstream-analysis-of-the-loadings">
<h2>Downstream analysis of the loadings<a class="headerlink" href="#Downstream-analysis-of-the-loadings" title="Permalink to this headline"></a></h2>
<p>We will now explore what molecular features Factor 14 and Factor 15 capture.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># specify the view of which we want to retrieve the loadings
# the loading matrix has dimensions factors x features
W_rna = sofa.tl.get_loadings(model, view=&quot;RNA&quot;)
W_rna
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>feature_name</th>
      <th>SHOX</th>
      <th>CSF2RA</th>
      <th>P2RY8</th>
      <th>CD99</th>
      <th>XG</th>
      <th>GYG2</th>
      <th>ARSF</th>
      <th>MXRA5</th>
      <th>PRKX</th>
      <th>STS</th>
      <th>...</th>
      <th>MX2</th>
      <th>TMPRSS2</th>
      <th>TMPRSS3</th>
      <th>UBASH3A</th>
      <th>TRPM2</th>
      <th>TSPEAR</th>
      <th>KRTAP12-3</th>
      <th>ITGB2</th>
      <th>COL18A1</th>
      <th>COL6A2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Factor_1 (astrocyte)</th>
      <td>-0.018227</td>
      <td>0.016628</td>
      <td>0.008709</td>
      <td>-0.085218</td>
      <td>-0.032678</td>
      <td>-0.051249</td>
      <td>-0.128016</td>
      <td>-0.017553</td>
      <td>0.020427</td>
      <td>0.006717</td>
      <td>...</td>
      <td>-0.013714</td>
      <td>-0.118789</td>
      <td>-0.051786</td>
      <td>-0.068351</td>
      <td>-0.008388</td>
      <td>-0.070681</td>
      <td>-0.040730</td>
      <td>-0.034352</td>
      <td>-0.009784</td>
      <td>-0.041832</td>
    </tr>
    <tr>
      <th>Factor_2 (caudal ganglionic eminence derived interneuron)</th>
      <td>0.061711</td>
      <td>0.025978</td>
      <td>0.041309</td>
      <td>0.008997</td>
      <td>0.103602</td>
      <td>0.066791</td>
      <td>0.096115</td>
      <td>0.026817</td>
      <td>0.018299</td>
      <td>0.073380</td>
      <td>...</td>
      <td>0.063504</td>
      <td>0.054544</td>
      <td>0.064555</td>
      <td>0.087258</td>
      <td>0.127651</td>
      <td>0.140096</td>
      <td>0.034622</td>
      <td>0.046414</td>
      <td>0.055640</td>
      <td>0.089154</td>
    </tr>
    <tr>
      <th>Factor_3 (endothelial cell)</th>
      <td>0.002676</td>
      <td>0.046630</td>
      <td>-0.063564</td>
      <td>0.013813</td>
      <td>-0.022183</td>
      <td>0.019438</td>
      <td>0.019055</td>
      <td>-0.016783</td>
      <td>-0.012183</td>
      <td>-0.014433</td>
      <td>...</td>
      <td>-0.028807</td>
      <td>0.011751</td>
      <td>-0.020785</td>
      <td>-0.045028</td>
      <td>0.020811</td>
      <td>-0.038803</td>
      <td>-0.009141</td>
      <td>0.003219</td>
      <td>-0.052285</td>
      <td>-0.009642</td>
    </tr>
    <tr>
      <th>Factor_4 (glutamatergic neuron)</th>
      <td>0.044140</td>
      <td>-0.032026</td>
      <td>-0.020472</td>
      <td>-0.004024</td>
      <td>0.047451</td>
      <td>0.072089</td>
      <td>0.020697</td>
      <td>0.033083</td>
      <td>-0.000998</td>
      <td>0.157253</td>
      <td>...</td>
      <td>0.066016</td>
      <td>0.010713</td>
      <td>0.008618</td>
      <td>0.068955</td>
      <td>0.070585</td>
      <td>0.160543</td>
      <td>0.067635</td>
      <td>-0.021258</td>
      <td>-0.048109</td>
      <td>0.078915</td>
    </tr>
    <tr>
      <th>Factor_5 (inhibitory interneuron)</th>
      <td>0.021449</td>
      <td>0.076019</td>
      <td>0.071614</td>
      <td>0.111477</td>
      <td>0.054973</td>
      <td>0.011741</td>
      <td>-0.012790</td>
      <td>-0.002646</td>
      <td>-0.000420</td>
      <td>0.029649</td>
      <td>...</td>
      <td>0.103795</td>
      <td>0.108892</td>
      <td>0.043622</td>
      <td>0.085356</td>
      <td>0.144796</td>
      <td>0.155512</td>
      <td>0.069746</td>
      <td>0.102761</td>
      <td>0.190089</td>
      <td>0.099793</td>
    </tr>
    <tr>
      <th>Factor_6 (medial ganglionic eminence derived interneuron)</th>
      <td>-0.002214</td>
      <td>0.011589</td>
      <td>0.009022</td>
      <td>0.005203</td>
      <td>-0.001285</td>
      <td>0.022863</td>
      <td>0.043369</td>
      <td>-0.009789</td>
      <td>-0.055886</td>
      <td>-0.051018</td>
      <td>...</td>
      <td>-0.000926</td>
      <td>0.001854</td>
      <td>0.013125</td>
      <td>-0.046666</td>
      <td>-0.175756</td>
      <td>-0.010585</td>
      <td>0.021193</td>
      <td>0.029928</td>
      <td>0.040288</td>
      <td>-0.004638</td>
    </tr>
    <tr>
      <th>Factor_7 (microglial cell)</th>
      <td>0.033687</td>
      <td>-0.371200</td>
      <td>-0.201232</td>
      <td>-0.180759</td>
      <td>0.048073</td>
      <td>0.049734</td>
      <td>0.059747</td>
      <td>0.044436</td>
      <td>-0.062388</td>
      <td>0.013768</td>
      <td>...</td>
      <td>-0.146758</td>
      <td>0.037564</td>
      <td>0.033068</td>
      <td>0.012605</td>
      <td>-0.099360</td>
      <td>0.029247</td>
      <td>0.020464</td>
      <td>-0.306571</td>
      <td>-0.021651</td>
      <td>0.055954</td>
    </tr>
    <tr>
      <th>Factor_8 (neural progenitor cell)</th>
      <td>-0.040064</td>
      <td>0.018456</td>
      <td>-0.022026</td>
      <td>0.028713</td>
      <td>0.042819</td>
      <td>-0.024957</td>
      <td>-0.036244</td>
      <td>-0.004213</td>
      <td>-0.068642</td>
      <td>-0.052252</td>
      <td>...</td>
      <td>0.021161</td>
      <td>0.001159</td>
      <td>0.014907</td>
      <td>0.055635</td>
      <td>0.011421</td>
      <td>0.031352</td>
      <td>0.032900</td>
      <td>0.015596</td>
      <td>-0.012561</td>
      <td>0.020977</td>
    </tr>
    <tr>
      <th>Factor_9 (oligodendrocyte)</th>
      <td>-0.045553</td>
      <td>-0.086439</td>
      <td>-0.095924</td>
      <td>-0.091932</td>
      <td>-0.071380</td>
      <td>-0.057834</td>
      <td>-0.062374</td>
      <td>-0.035606</td>
      <td>-0.074925</td>
      <td>-0.039691</td>
      <td>...</td>
      <td>-0.082786</td>
      <td>-0.074244</td>
      <td>-0.033555</td>
      <td>-0.044907</td>
      <td>-0.114517</td>
      <td>-0.094449</td>
      <td>-0.031309</td>
      <td>-0.104703</td>
      <td>-0.005420</td>
      <td>-0.073570</td>
    </tr>
    <tr>
      <th>Factor_10 (oligodendrocyte precursor cell)</th>
      <td>0.013328</td>
      <td>-0.076632</td>
      <td>-0.095766</td>
      <td>-0.061523</td>
      <td>0.004489</td>
      <td>-0.009986</td>
      <td>-0.035331</td>
      <td>-0.014284</td>
      <td>-0.010431</td>
      <td>-0.086027</td>
      <td>...</td>
      <td>-0.005166</td>
      <td>0.103091</td>
      <td>0.000971</td>
      <td>-0.002350</td>
      <td>-0.054968</td>
      <td>-0.038637</td>
      <td>-0.016010</td>
      <td>-0.023825</td>
      <td>0.012225</td>
      <td>-0.018629</td>
    </tr>
    <tr>
      <th>Factor_11 (pericyte)</th>
      <td>-0.068400</td>
      <td>-0.117113</td>
      <td>-0.107469</td>
      <td>-0.063415</td>
      <td>-0.119831</td>
      <td>-0.101984</td>
      <td>-0.116682</td>
      <td>-0.038292</td>
      <td>-0.105895</td>
      <td>-0.142456</td>
      <td>...</td>
      <td>-0.074856</td>
      <td>-0.127596</td>
      <td>-0.068352</td>
      <td>-0.096209</td>
      <td>-0.098300</td>
      <td>-0.039917</td>
      <td>0.005252</td>
      <td>-0.071315</td>
      <td>0.125981</td>
      <td>0.006964</td>
    </tr>
    <tr>
      <th>Factor_12 (radial glial cell)</th>
      <td>0.013457</td>
      <td>0.035862</td>
      <td>0.016619</td>
      <td>-0.008517</td>
      <td>-0.005948</td>
      <td>-0.048667</td>
      <td>-0.013628</td>
      <td>-0.006250</td>
      <td>0.036617</td>
      <td>0.091800</td>
      <td>...</td>
      <td>0.037933</td>
      <td>0.001341</td>
      <td>0.018385</td>
      <td>0.011399</td>
      <td>0.078934</td>
      <td>0.081725</td>
      <td>0.034904</td>
      <td>0.064347</td>
      <td>0.076669</td>
      <td>0.018390</td>
    </tr>
    <tr>
      <th>Factor_13 (vascular associated smooth muscle cell)</th>
      <td>-0.020864</td>
      <td>-0.006278</td>
      <td>0.057538</td>
      <td>0.021366</td>
      <td>0.016715</td>
      <td>-0.011897</td>
      <td>0.036587</td>
      <td>-0.026533</td>
      <td>0.022503</td>
      <td>0.039456</td>
      <td>...</td>
      <td>0.015130</td>
      <td>-0.010799</td>
      <td>0.014685</td>
      <td>0.013097</td>
      <td>-0.001941</td>
      <td>0.011595</td>
      <td>0.008483</td>
      <td>0.012239</td>
      <td>-0.133319</td>
      <td>-0.082095</td>
    </tr>
    <tr>
      <th>Factor_14</th>
      <td>-0.030205</td>
      <td>-0.065338</td>
      <td>-0.098833</td>
      <td>-0.098750</td>
      <td>-0.020684</td>
      <td>0.015680</td>
      <td>0.004504</td>
      <td>0.002585</td>
      <td>-0.049090</td>
      <td>0.002668</td>
      <td>...</td>
      <td>-0.099198</td>
      <td>-0.061864</td>
      <td>-0.059691</td>
      <td>-0.070308</td>
      <td>-0.167516</td>
      <td>-0.050790</td>
      <td>-0.045526</td>
      <td>-0.186547</td>
      <td>0.030531</td>
      <td>-0.045366</td>
    </tr>
    <tr>
      <th>Factor_15</th>
      <td>0.102394</td>
      <td>0.111089</td>
      <td>0.161223</td>
      <td>0.145873</td>
      <td>0.143792</td>
      <td>0.096924</td>
      <td>0.109016</td>
      <td>0.141172</td>
      <td>0.169392</td>
      <td>0.152441</td>
      <td>...</td>
      <td>0.147098</td>
      <td>0.099834</td>
      <td>0.066984</td>
      <td>0.102654</td>
      <td>0.077644</td>
      <td>0.146752</td>
      <td>0.049644</td>
      <td>0.127989</td>
      <td>0.083689</td>
      <td>0.177638</td>
    </tr>
  </tbody>
</table>
<p>15 rows × 2000 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>W_atac = sofa.tl.get_loadings(model, view=&quot;ATAC&quot;)
W_atac
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>feature_name</th>
      <th>HES5</th>
      <th>PRDM16</th>
      <th>LINC01134</th>
      <th>SLC2A5</th>
      <th>PIK3CD</th>
      <th>TNFRSF1B</th>
      <th>AADACL4</th>
      <th>SLC25A34-AS1</th>
      <th>PADI2</th>
      <th>PADI1</th>
      <th>...</th>
      <th>LINC00279</th>
      <th>MT-ND1</th>
      <th>MT-ND2</th>
      <th>MT-CO1</th>
      <th>MT-CO2</th>
      <th>MT-ATP6</th>
      <th>MT-ND3</th>
      <th>MT-ND4L</th>
      <th>MT-ND4</th>
      <th>MT-ND5</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Factor_1 (astrocyte)</th>
      <td>-0.208182</td>
      <td>-0.429047</td>
      <td>-0.000131</td>
      <td>0.045786</td>
      <td>0.080833</td>
      <td>0.054359</td>
      <td>-0.012543</td>
      <td>-0.025898</td>
      <td>-0.091752</td>
      <td>0.003296</td>
      <td>...</td>
      <td>0.036371</td>
      <td>-0.049887</td>
      <td>-0.039553</td>
      <td>0.012043</td>
      <td>-0.031498</td>
      <td>-0.078227</td>
      <td>-0.136771</td>
      <td>-0.020041</td>
      <td>-0.042751</td>
      <td>-0.024138</td>
    </tr>
    <tr>
      <th>Factor_2 (caudal ganglionic eminence derived interneuron)</th>
      <td>-0.036767</td>
      <td>-0.069018</td>
      <td>0.001598</td>
      <td>-0.011377</td>
      <td>-0.025387</td>
      <td>-0.038779</td>
      <td>0.004101</td>
      <td>0.004330</td>
      <td>-0.035379</td>
      <td>0.003445</td>
      <td>...</td>
      <td>0.001888</td>
      <td>0.138144</td>
      <td>0.123199</td>
      <td>0.152745</td>
      <td>0.134151</td>
      <td>0.095955</td>
      <td>0.093427</td>
      <td>0.094417</td>
      <td>0.135157</td>
      <td>0.092551</td>
    </tr>
    <tr>
      <th>Factor_3 (endothelial cell)</th>
      <td>0.026845</td>
      <td>0.086176</td>
      <td>-0.006730</td>
      <td>0.032492</td>
      <td>0.092398</td>
      <td>-0.062162</td>
      <td>0.009899</td>
      <td>0.019996</td>
      <td>0.012182</td>
      <td>0.006935</td>
      <td>...</td>
      <td>0.010711</td>
      <td>0.007029</td>
      <td>-0.003770</td>
      <td>0.000740</td>
      <td>0.000596</td>
      <td>-0.013372</td>
      <td>-0.011830</td>
      <td>0.009384</td>
      <td>-0.013850</td>
      <td>-0.012648</td>
    </tr>
    <tr>
      <th>Factor_4 (glutamatergic neuron)</th>
      <td>-0.119810</td>
      <td>-0.110515</td>
      <td>-0.008428</td>
      <td>-0.065243</td>
      <td>-0.076141</td>
      <td>-0.082045</td>
      <td>0.001842</td>
      <td>-0.012405</td>
      <td>-0.123653</td>
      <td>-0.004981</td>
      <td>...</td>
      <td>-0.027927</td>
      <td>-0.064613</td>
      <td>-0.099515</td>
      <td>-0.053974</td>
      <td>-0.108447</td>
      <td>-0.121445</td>
      <td>-0.160079</td>
      <td>-0.058071</td>
      <td>-0.102936</td>
      <td>-0.070332</td>
    </tr>
    <tr>
      <th>Factor_5 (inhibitory interneuron)</th>
      <td>0.074077</td>
      <td>0.080980</td>
      <td>-0.000610</td>
      <td>0.054926</td>
      <td>0.059176</td>
      <td>0.070569</td>
      <td>-0.001754</td>
      <td>-0.002609</td>
      <td>0.111904</td>
      <td>0.008717</td>
      <td>...</td>
      <td>0.032175</td>
      <td>0.144616</td>
      <td>0.140038</td>
      <td>0.144284</td>
      <td>0.186204</td>
      <td>0.150951</td>
      <td>0.152748</td>
      <td>0.052082</td>
      <td>0.145854</td>
      <td>0.084795</td>
    </tr>
    <tr>
      <th>Factor_6 (medial ganglionic eminence derived interneuron)</th>
      <td>0.061534</td>
      <td>0.110697</td>
      <td>-0.012850</td>
      <td>0.014597</td>
      <td>-0.108322</td>
      <td>0.006335</td>
      <td>-0.001688</td>
      <td>0.008390</td>
      <td>0.036388</td>
      <td>-0.003137</td>
      <td>...</td>
      <td>0.001910</td>
      <td>-0.249092</td>
      <td>-0.228440</td>
      <td>-0.332192</td>
      <td>-0.354922</td>
      <td>-0.283303</td>
      <td>-0.202077</td>
      <td>-0.081255</td>
      <td>-0.252142</td>
      <td>-0.182548</td>
    </tr>
    <tr>
      <th>Factor_7 (microglial cell)</th>
      <td>0.073273</td>
      <td>0.076353</td>
      <td>-0.015753</td>
      <td>-0.372691</td>
      <td>-0.235887</td>
      <td>-0.319122</td>
      <td>-0.010101</td>
      <td>-0.001108</td>
      <td>-0.065391</td>
      <td>-0.022516</td>
      <td>...</td>
      <td>0.011044</td>
      <td>0.019545</td>
      <td>0.015750</td>
      <td>0.001131</td>
      <td>0.000565</td>
      <td>-0.025114</td>
      <td>-0.020271</td>
      <td>0.019034</td>
      <td>0.008098</td>
      <td>0.033560</td>
    </tr>
    <tr>
      <th>Factor_8 (neural progenitor cell)</th>
      <td>0.094023</td>
      <td>-0.073044</td>
      <td>-0.020015</td>
      <td>-0.010783</td>
      <td>0.006819</td>
      <td>-0.035352</td>
      <td>-0.022101</td>
      <td>0.004570</td>
      <td>0.003941</td>
      <td>0.013296</td>
      <td>...</td>
      <td>0.004650</td>
      <td>0.018530</td>
      <td>0.006067</td>
      <td>0.097694</td>
      <td>0.021061</td>
      <td>-0.002756</td>
      <td>-0.011136</td>
      <td>-0.008039</td>
      <td>0.041842</td>
      <td>0.008941</td>
    </tr>
    <tr>
      <th>Factor_9 (oligodendrocyte)</th>
      <td>-0.096187</td>
      <td>-0.090426</td>
      <td>0.002811</td>
      <td>-0.057448</td>
      <td>-0.100601</td>
      <td>-0.064315</td>
      <td>0.001830</td>
      <td>0.006622</td>
      <td>0.256400</td>
      <td>0.001876</td>
      <td>...</td>
      <td>0.140956</td>
      <td>-0.023779</td>
      <td>0.035511</td>
      <td>0.045417</td>
      <td>0.096452</td>
      <td>0.008418</td>
      <td>0.016682</td>
      <td>-0.010833</td>
      <td>0.055347</td>
      <td>-0.018010</td>
    </tr>
    <tr>
      <th>Factor_10 (oligodendrocyte precursor cell)</th>
      <td>0.109106</td>
      <td>-0.139850</td>
      <td>-0.001256</td>
      <td>-0.085660</td>
      <td>-0.103560</td>
      <td>-0.073114</td>
      <td>-0.011707</td>
      <td>-0.001273</td>
      <td>-0.034249</td>
      <td>0.001216</td>
      <td>...</td>
      <td>-0.023487</td>
      <td>-0.091286</td>
      <td>-0.058322</td>
      <td>-0.080787</td>
      <td>-0.033584</td>
      <td>-0.034352</td>
      <td>-0.039314</td>
      <td>-0.031082</td>
      <td>-0.040277</td>
      <td>-0.050600</td>
    </tr>
    <tr>
      <th>Factor_11 (pericyte)</th>
      <td>-0.096806</td>
      <td>-0.082905</td>
      <td>-0.013746</td>
      <td>-0.031378</td>
      <td>0.182597</td>
      <td>0.024180</td>
      <td>0.004992</td>
      <td>0.009840</td>
      <td>-0.045573</td>
      <td>0.001019</td>
      <td>...</td>
      <td>-0.022335</td>
      <td>0.069674</td>
      <td>0.102980</td>
      <td>0.051257</td>
      <td>0.088180</td>
      <td>0.073444</td>
      <td>0.077907</td>
      <td>0.039346</td>
      <td>0.044237</td>
      <td>0.039888</td>
    </tr>
    <tr>
      <th>Factor_12 (radial glial cell)</th>
      <td>-0.305815</td>
      <td>-0.370938</td>
      <td>0.014988</td>
      <td>0.027542</td>
      <td>0.026361</td>
      <td>0.024751</td>
      <td>-0.001839</td>
      <td>0.037604</td>
      <td>0.079338</td>
      <td>0.000370</td>
      <td>...</td>
      <td>0.007548</td>
      <td>0.139923</td>
      <td>0.137401</td>
      <td>0.122713</td>
      <td>0.170129</td>
      <td>0.160537</td>
      <td>0.122313</td>
      <td>0.048193</td>
      <td>0.146312</td>
      <td>0.089201</td>
    </tr>
    <tr>
      <th>Factor_13 (vascular associated smooth muscle cell)</th>
      <td>0.097635</td>
      <td>0.021009</td>
      <td>-0.025192</td>
      <td>0.014730</td>
      <td>0.082244</td>
      <td>0.057195</td>
      <td>-0.006985</td>
      <td>-0.018558</td>
      <td>0.041190</td>
      <td>-0.010776</td>
      <td>...</td>
      <td>0.011489</td>
      <td>0.117726</td>
      <td>0.129374</td>
      <td>0.112941</td>
      <td>0.148383</td>
      <td>0.155056</td>
      <td>0.177456</td>
      <td>0.069603</td>
      <td>0.138858</td>
      <td>0.053408</td>
    </tr>
    <tr>
      <th>Factor_14</th>
      <td>0.064157</td>
      <td>-0.044716</td>
      <td>-0.024030</td>
      <td>0.007258</td>
      <td>0.054465</td>
      <td>-0.057635</td>
      <td>0.013054</td>
      <td>-0.013529</td>
      <td>0.020315</td>
      <td>0.006386</td>
      <td>...</td>
      <td>0.040675</td>
      <td>-0.076650</td>
      <td>-0.030133</td>
      <td>-0.004744</td>
      <td>-0.037376</td>
      <td>-0.099526</td>
      <td>-0.079796</td>
      <td>0.008935</td>
      <td>-0.054390</td>
      <td>-0.010699</td>
    </tr>
    <tr>
      <th>Factor_15</th>
      <td>0.004490</td>
      <td>-0.062551</td>
      <td>0.004475</td>
      <td>0.026869</td>
      <td>0.085769</td>
      <td>0.083807</td>
      <td>0.021577</td>
      <td>-0.018813</td>
      <td>-0.007668</td>
      <td>-0.005300</td>
      <td>...</td>
      <td>-0.029235</td>
      <td>0.010413</td>
      <td>0.031491</td>
      <td>0.010445</td>
      <td>0.000602</td>
      <td>0.039254</td>
      <td>0.070796</td>
      <td>-0.004739</td>
      <td>0.025249</td>
      <td>-0.005580</td>
    </tr>
  </tbody>
</table>
<p>15 rows × 2000 columns</p>
</div></div>
</div>
<p>To plot the top loadings use:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># plot the top negative RNA loadings of Factor 7 (microglial cells)
sofa.pl.plot_top_loadings(model, view=&quot;RNA&quot;, factor = 7, top_n=20, sign=&quot;-&quot;)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Axes: xlabel=&#39;Loadings&#39;, ylabel=&#39;Features&#39;&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_sc_multiome_example_50_1.png" src="../_images/notebooks_sc_multiome_example_50_1.png" />
</div>
</div>
<p>Among the top loadings of Factor 7 (guided by microglial cell label) are markers for microglial cells (CSF1R, CX3CR1, TYROBP).</p>
<section id="Gene-set-overrepresentation-analysis">
<h3>Gene set overrepresentation analysis<a class="headerlink" href="#Gene-set-overrepresentation-analysis" title="Permalink to this headline"></a></h3>
<p>We can use gene set overrepresentation analysis on the top loadings of a given factor. Factors typically represent axes of variation having a positive and a negative end. We will therefore both investigate the positive or negative loadings. This gives us an overview of which gene sets are enriched in this factor and what biological processes it probably captured.</p>
<section id="Factor-14">
<h4>Factor 14<a class="headerlink" href="#Factor-14" title="Permalink to this headline"></a></h4>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># get top 100 positive loadings of ATAC view
loadings = sofa.tl.get_top_loadings(model,factor=14, view=&quot;ATAC&quot;, sign=&quot;+&quot;, top_n=100)


sofa.pl.plot_enrichment(loadings,
                        background=Xmdata.mod[&quot;ATAC&quot;].var, # all genes considered in the analysis, used as background
                        db=[ &quot;GO_Biological_Process_2023&quot;],  # a list of databases for overrepresentation analysis,
                        top_n=[8]) # the number of genesets for each database to plot
# sofa.pl.plot_enrichment uses the enrichr API, please refer to https://maayanlab.cloud/Enrichr/#libraries for a full list of available databases
<br/></pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Axes: xlabel=&#39;-log10 adjusted p-values&#39;, ylabel=&#39;Gene sets&#39;&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_sc_multiome_example_54_1.png" src="../_images/notebooks_sc_multiome_example_54_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># get top 100 negative loadings of ATAC view
loadings = sofa.tl.get_top_loadings(model,factor=14, view=&quot;ATAC&quot;, sign=&quot;-&quot;, top_n=100)


sofa.pl.plot_enrichment(loadings,
                        background=Xmdata.mod[&quot;ATAC&quot;].var, # all genes considered in the analysis, used as background
                        db=[&quot;GO_Biological_Process_2023&quot;],  # a list of databases for overrepresentation analysis,
                        top_n=[8]) # the number of genesets for each database to plot
# sofa.pl.plot_enrichment uses the enrichr API, please refer to https://maayanlab.cloud/Enrichr/#libraries for a full list of available databases
<br/></pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Axes: xlabel=&#39;-log10 adjusted p-values&#39;, ylabel=&#39;Gene sets&#39;&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_sc_multiome_example_55_1.png" src="../_images/notebooks_sc_multiome_example_55_1.png" />
</div>
</div>
</section>
<section id="Factor-15">
<h4>Factor 15<a class="headerlink" href="#Factor-15" title="Permalink to this headline"></a></h4>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># get top 100 positive loadings of RNA view
loadings = sofa.tl.get_top_loadings(model,factor=15, view=&quot;RNA&quot;, sign=&quot;+&quot;, top_n=100)


sofa.pl.plot_enrichment(loadings,
                        background=Xmdata.mod[&quot;RNA&quot;].var, # all genes considered in the analysis, used as background
                        db=[&quot;GO_Biological_Process_2023&quot;],  # a list of databases for overrepresentation analysis,
                        top_n=[8]) # the number of genesets for each database to plot
# sofa.pl.plot_enrichment uses the enrichr API, please refer to https://maayanlab.cloud/Enrichr/#libraries for a full list of available databases
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Axes: xlabel=&#39;-log10 adjusted p-values&#39;, ylabel=&#39;Gene sets&#39;&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_sc_multiome_example_57_1.png" src="../_images/notebooks_sc_multiome_example_57_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># get top 100 negative loadings of RNA view
loadings = sofa.tl.get_top_loadings(model,factor=15, view=&quot;RNA&quot;, sign=&quot;-&quot;, top_n=100)


sofa.pl.plot_enrichment(loadings,
                        background=Xmdata.mod[&quot;RNA&quot;].var, # all genes considered in the analysis, used as background
                        db=[&quot;GO_Biological_Process_2023&quot;],  # a list of databases for overrepresentation analysis,
                        top_n=[8]) # the number of genesets for each database to plot
# sofa.pl.plot_enrichment uses the enrichr API, please refer to https://maayanlab.cloud/Enrichr/#libraries for a full list of available databases
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Axes: xlabel=&#39;-log10 adjusted p-values&#39;, ylabel=&#39;Gene sets&#39;&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_sc_multiome_example_58_1.png" src="../_images/notebooks_sc_multiome_example_58_1.png" />
</div>
</div>
</section>
</section>
</section>
</section>


           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Tuemay Capraz.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>